ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerViewModel.kt

```diff
@@ -89,61 +87,76 @@ internal constructor(
+  private val _suggestLoiJobs: MutableStateFlow<List<Job>> = MutableStateFlow(listOf())
+  val suggestLoiJobs: StateFlow<List<Job>> =
+    _suggestLoiJobs.stateIn(viewModelScope, SharingStarted.Lazily, listOf())
 
   init {
     // THIS SHOULD NOT BE CALLED ON CONFIG CHANGE
     // TODO: Clear location of interest markers when survey is deactivated.
     // TODO: Since we depend on survey stream from repo anyway, this transformation can be moved
     //  into the repository.
 
-    // LOIs that are persisted to the local and remote dbs.
-    mapLocationOfInterestFeatures =
-      surveyRepository.activeSurveyFlow
-        .transform { survey ->
-          if (survey == null) {
-            emit(setOf())
-          } else {
-            emitAll(locationOfInterestRepository.findLocationsOfInterestFeatures(survey))
-          }
-        }
-        .distinctUntilChanged()
-        .stateIn(viewModelScope, SharingStarted.Lazily, setOf())
-
-    loisWithinMapBoundsAtVisibleZoomLevel =
-      surveyRepository.activeSurveyFlowable
-        .switchMap { survey ->
-          cameraZoomUpdates.switchMap { zoomLevel ->
-            if (zoomLevel >= CLUSTERING_ZOOM_THRESHOLD && survey.isPresent)
-              locationOfInterestRepository.getWithinBoundsOnceAndStream(
-                survey.get(),
-                cameraBoundUpdates
-              )
-            else Flowable.just(listOf())
-          }
-        }
-        .toLiveData()
-
-    suggestLoiJobs =
-      surveyRepository.activeSurveyFlow
-        .combine(cameraZoomUpdates.asFlow()) { survey, zoomLevel ->
-          if (zoomLevel < CLUSTERING_ZOOM_THRESHOLD) {
-            listOf()
-          } else {
-            survey?.jobs?.filter { job -> job.suggestLoiTaskType != null }?.toList() ?: listOf()
-          }
-        }
-        .distinctUntilChanged()
+    viewModelScope.launch {
+      surveyRepository.activeSurveyFlow.collect {
+        _survey = it
+        refreshMapFeaturesAndCards(it)
+      }
+    }
+  }
+
+  private suspend fun refreshMapFeaturesAndCards(survey: Survey?) {
+    updateMapFeatures(survey)
+    updateLoisAndJobs(survey, currentCameraPosition)
+  }
+
+  private suspend fun updateLoisAndJobs(survey: Survey?, cameraPosition: CameraPosition?) {
+    updateMapLois(survey, cameraPosition)
+    updateSuggestLoiJobs(survey, cameraPosition)
+  }
+
+  private suspend fun updateMapFeatures(survey: Survey?) {
+    if (survey == null) {
+      _mapLoiFeatures.value = setOf()
+    } else {
+      // LOIs that are persisted to the local and remote dbs.
+      _mapLoiFeatures.emitAll(locationOfInterestRepository.findLocationsOfInterestFeatures(survey))
+    }
+  }
+
+  private suspend fun updateMapLois(survey: Survey?, cameraPosition: CameraPosition?) {
+    val bounds = cameraPosition?.bounds
+    val isZoomLevelLessThanThreshold = cameraPosition.isLessThanThreshold(CLUSTERING_ZOOM_THRESHOLD)
+    if (bounds == null || survey == null || isZoomLevelLessThanThreshold) {
```

Comment:
Not sure the added indirection of the helper here adds clarity. 

As an alternative, you could get rid of `val isZoomLevelLessThanThreshold`, rename the helper `isBelowClusteringZoomThreshold` and move the constant into the helper. Then you'd end up with:

```kt
   val bounds = cameraPosition?.bounds
    if (bounds == null || survey == null || cameraPosition.isBelowZoomThreshold()) {
   ...
    }
```

Or similar.

Reply:
Done